ambry-messageformat/src/main/java/com.github.ambry.messageformat/MessageSievingInputStream.java

```diff
@@ -152,81 +184,153 @@ public boolean hasInvalidMessages() {
+    return sievedMessageInfoList;
   }
 
   /**
-   * Ensures blob validity in the given input stream. For now, blobs are checked for message corruption
-   * @param byteArrayInputStream stream against which validation has to be done
-   * @param size total size of the message expected
-   * @param currentOffset Current offset at which the data has to be read from the given byte array
-   * @param storeKeyFactory StoreKeyFactory used to get store key
-   * @return true if message is valid and false otherwise
-   * @throws IOException
+   * Validates and potentially transforms the given input stream consisting of message data. It does so using the list
+   * of {@link Transformer}s associated with this instance.
+   * message corruption and acceptable formats.
+   * @param inMsg the original {@link Message} that needs to be validated and possibly transformed.
+   * @param sievedOutputStream the output {@link OutputStream}
+   * @param msgOffset the offset of the message in the stream.
+   * @throws Exception if an exception was encountered by any of the transformers.
    */
-  private boolean checkForMessageValidity(ByteArrayInputStream byteArrayInputStream, int currentOffset, long size,
-      StoreKeyFactory storeKeyFactory, MessageInfo msgInfo) throws IOException {
-    boolean isValid = false;
-    ByteBuffer encryptionKey;
-    BlobProperties props;
-    ByteBuffer metadata;
-    BlobData blobData;
-    long startTime = SystemTime.getInstance().milliseconds();
-    try {
-      int availableBeforeParsing = byteArrayInputStream.available();
-      byte[] headerVersionInBytes = new byte[Version_Field_Size_In_Bytes];
-      byteArrayInputStream.read(headerVersionInBytes, 0, Version_Field_Size_In_Bytes);
-      ByteBuffer headerVersion = ByteBuffer.wrap(headerVersionInBytes);
-      short version = headerVersion.getShort();
-      if (!isValidHeaderVersion(version)) {
-        throw new MessageFormatException("Header version not supported " + version,
-            MessageFormatErrorCodes.Data_Corrupt);
-      }
-      ByteBuffer headerBuffer = ByteBuffer.allocate(getHeaderSizeForVersion(version));
-      headerBuffer.putShort(version);
-      byteArrayInputStream.read(headerBuffer.array(), 2, headerBuffer.capacity() - 2);
-      headerBuffer.position(headerBuffer.capacity());
-      headerBuffer.flip();
-      MessageHeader_Format header = getMessageHeader(version, headerBuffer);
-      StoreKey storeKey = storeKeyFactory.getStoreKey(new DataInputStream(byteArrayInputStream));
-      if (header.isPutRecord()) {
-        encryptionKey = header.hasEncryptionKeyRecord() ? deserializeBlobEncryptionKey(byteArrayInputStream) : null;
-        props = deserializeBlobProperties(byteArrayInputStream);
-        metadata = deserializeUserMetadata(byteArrayInputStream);
-        blobData = deserializeBlob(byteArrayInputStream);
+  private void sieve(Message inMsg, OutputStream sievedOutputStream, int msgOffset) throws Exception {
```

Comment:
What exception does this method throw?  The transform methods all stuff their exceptions in the return value.

Reply:
Transformers are pluggable, and the interface allows for `Exception` to be thrown, so that had to be handled here.

Reply:
Ok.